src/System.CommandLine.Subsystems/ValueProvider.cs

```diff
@@ -37,29 +38,53 @@ private bool TryGetValue<T>(CliSymbol symbol, out T? value)
     public T? GetValue<T>(CliValueSymbol valueSymbol)
         => GetValueInternal<T>(valueSymbol);
 
-    private T? GetValueInternal<T>(CliValueSymbol? valueSymbol)
+    private T? GetValueInternal<T>(CliValueSymbol valueSymbol)
     {
+        var _ = valueSymbol ?? throw new ArgumentNullException(nameof(valueSymbol));
```

Comment:
Provide guard to protect against circularity. 

For example, mark the cache as currently being calculated